### ISSUE TITLE
Nvidia GPU not visible via CUDA backend in plaidml-setup

### ISSUE THREAD

**Issue by gyenesvi** [plaidml/plaidml#176]
Hi,

I am wondering if the CUDA backend is fully implemented already or is it yet in an experimental phase? I can see it has been added lately, but I cannot get plaidml use it. After installing plaidml, plaidml-setup lists the GPU only as a device using the OpenCL backend: opencl_nvidia_geforce_gtx_titan_x (plus I see llvm_cpu device). I guess I should see the same GPU device using cuda backend as well, and I should be able to set it as the target device, right? I tried this on several machines with Nvidia GPUs and I always only get the devices with OpenCL backend. Am I missing some setup here, or is this feature not fully functional yet?

**Comment by brianretford** [comment]
Currently you have to build it yourself to enable CUDA, against whatever version of CUDA you have. In our experience performance is identical to OpenCL, and distribution with CUDA is complicated. If you set VAI_NEED_CUDA to true before building it should enable it.

**Comment by brianretford** [comment]
No we haven’t really but that’s a really good point. It wouldn’t be that hard to add support for initializing it with custom contexts. The CUDA devices should definitely be there if you rebuild it – you build the plaidml wheel and installed that?

**Comment by gyenesvi** [comment]
Yes, rebuilt the wheel and installed that. I'll try again, currently I have no access to that machine.

I guess the shared contexts should be applicable in case of OpenCL/Metal too. Shall I open a separate issue for that?

**Comment by gyenesvi** [comment]
I tried rebuilding and installing from source. I enabled cuda with export VAI_NEED_CUDA=1, and then the compilation was definitely different. Without cuda enabled it compiled fine, however, with cuda enabled, there were compilation errors, resulting from inconsistencies among the files tile/hal/cuda/hal.h and tile/hal/cuda/hal.cc, which I resolved as follows:

tile/hal/cuda/hal.h line 183
std::unique_ptr<hal::Kernel> -> std::shared_ptr<Kernel>
(pointer type and namespace is wrong)

tile/hal/cuda/compiler.cc line 116
CodeKernel -> ComputeKernel

tile/hal/cuda/hal.cc line 210, 212
I removed the param kidx

tile/hal/cuda/hal.cc line 225,226,227
Buffer -> hal::buffer
Event -> hal::Event
remove default value '= false' for param 'enable_profiling'

Furthermore, there was a compile error that resulted from the use of boost libraries. Both the standard library and boost define the function 'forward', albeit in different namespaces (std::forward and boost::forward). In the boost function 'make_ready_future' in boost/thread/future.hpp there is a call to 'forward' without a namespace prefix, and the compiler invoked in the bazel toolchain (seemingly gcc 4.8.5 in this case) complains that the call is ambiguous. I believe by the rules of C++ it should be unambiguous, unless there is a 'using namespace std' somewhere in your code (the function 'make_ready_future' is called several times in tile/hal/cuda/hal.cc). I was only able to get it compile by editing the boost code to use boost::forward (actually it should be that way, all other instances of 'forward' are used that way in boost), but it could be probably fixed by using a more up-to-date compiler, which I was unable to set in the bazel toolchain.

Anyway, after all of this, I was able to install it. However, the GPU is still not visible via cuda backend, only via the OpenCL backend.

**Comment by brianretford** [comment]
Frank did most of the work on the CUDA HAL and this is a bit beyond what I know. Frank, can you help Viktor out?

**Comment by colorfuldarkgray** [comment]
Hi

I wonder why trying to run plaidml with nvidia when the purpose of plaidml (I think) is to make use of other gpu than nvidia.

Best regards!

**Comment by brianretford** [comment]
That's a good question. We made the CUDA backend primarily so we could see if we were leaving any performance on the table relative OpenCL on nvidia cards. As to why you would use PlaidML on nvidia - there are some reasons:

a) no need to agree to the cudnn license which contains some absurd restrictions on what cards you can use for training in datacenters
b) some nvidia hardware that is still in production in the industrial space like the tk2 don't support modern cuDNN versions and thus aren't really useful for machine learning without plaid

**Comment by gyenesvi** [comment]
Another reason would be that if you have a system written in CUDA, and you want to use plaid for adding NN functionality, and the two needs to interoperate, like sharing buffers and passing data around. And of course rewriting the whole system in OpenCL is not an option.

**Comment by colorfuldarkgray** [comment]
@gyenesvi @brianretford 

Thank you!

**Comment by brianretford** [comment]
Yao, could you look into formally enabling CUDA support along with the other Nvidia work you're doing?
